nexus/src/authz/api_resources.rs

```diff
@@ -213,364 +186,110 @@ impl AuthorizedResource for Fleet {
+    roles_allowed = false,
+    polar_snippet = InProject,
 }
 
-impl Sled {
-    pub fn id(&self) -> Uuid {
-        self.sled_id
-    }
+authz_resource! {
+    name = "NetworkInterface",
+    parent = "Instance",
+    primary_key = Uuid,
+    roles_allowed = false,
+    polar_snippet = InProject,
 }
 
-impl oso::PolarClass for Sled {
-    fn get_polar_class_builder() -> oso::ClassBuilder<Self> {
-        oso::Class::builder()
-            .add_method(
-                "has_role",
-                // Roles are not supported on Sleds today.
-                |_: &Sled, _: AuthenticatedActor, _: String| false,
-            )
-            .add_attribute_getter("fleet", |_: &Sled| FLEET)
-    }
+authz_resource! {
+    name = "Vpc",
+    parent = "Project",
+    primary_key = Uuid,
+    roles_allowed = false,
+    polar_snippet = InProject,
 }
 
-impl ApiResource for Sled {
-    fn db_resource(&self) -> Option<(ResourceType, Uuid)> {
-        None
-    }
-
-    fn parent(&self) -> Option<&dyn AuthorizedResource> {
-        Some(&FLEET)
-    }
+authz_resource! {
+    name = "VpcRouter",
+    parent = "Vpc",
+    primary_key = Uuid,
+    roles_allowed = false,
+    polar_snippet = InProject,
 }
 
-impl ApiResourceError for Sled {
-    fn not_found(&self) -> Error {
-        self.lookup_type.clone().into_not_found(ResourceType::Sled)
-    }
+authz_resource! {
+    name = "RouterRoute",
+    parent = "VpcRouter",
+    primary_key = Uuid,
+    roles_allowed = false,
+    polar_snippet = InProject,
 }
 
-/// Represents a [`crate::db::model::Organization`] for authz purposes
-///
-/// This object is used for authorization checks on an Organization by passing
-/// it as the `resource` argument to [`crate::context::OpContext::authorize()`].
-/// You typically construct one of these with [`Fleet::organization()`].  You
-/// can perform authorization checks on children of the Organization (e.g.,
-/// Projects) using one of the methods below that return authz objects
-/// representing those children.
-#[derive(Clone, Debug)]
-pub struct Organization {
-    organization_id: Uuid,
-    lookup_type: LookupType,
+authz_resource! {
+    name = "VpcSubnet",
+    parent = "Vpc",
+    primary_key = Uuid,
+    roles_allowed = false,
+    polar_snippet = InProject,
 }
 
-impl Organization {
-    // TODO-cleanup It seems like we shouldn't need this.  It's currently used
-    // for compatibility so we don't have to change a ton of callers as we
-    // iterate on the lookup APIs.
-    pub fn id(&self) -> Uuid {
-        self.organization_id
-    }
+// Miscellaneous resources nested directly below "Fleet"
 
-    /// Returns an authz resource representing a child Project
-    pub fn project(
-        &self,
-        project_id: Uuid,
-        lookup_type: LookupType,
-    ) -> Project {
-        Project { parent: self.clone(), project_id, lookup_type }
-    }
+authz_resource! {
+    name = "Role",
+    parent = "Fleet",
+    primary_key = (String, String),
```

Comment:
It's too bad this couldn't be tied closer to what's in schema.rs, if something like
```
primary_key = db::schema::role_builtin::dsl::primary_key_type()
```
was possible, though even that statement is problematic because it'd probably return `sql_types::Text` instead of String.

Reply:
Agreed, that'd be a nice avenue for follow-up cleanup, if we can figure out how to do it.